File: Array_Functions/Reverse_an_Array/reverse_copy.cpp
#include <iostream>
#include <vector>

using namespace std;

template <typename T>
vector<T> reverse_copy(vector<T> &vc)
{
    int n = vc.size();
    vector<T> rev(n);
    for (int i = n - 1; i > -1; i--)
        rev[n - i - 1] = vc[i];

    return rev;
}

int main()
{
    vector<int> vc = {1, 2, 3};
    vector<int> rev = reverse_copy(vc);
    for (int i : rev)
        cout << i << " ";
    cout << endl;
    return 0;
}

// No builtin function to reverse a vector and return a copy

File: Array_Functions/Reverse an Array/reverse_copy.java
import java.util.ArrayList;
import java.util.List;

public class reverse_copy {
    public static List<T> reverseCopy(List<T> vc) {
        int n = vc.size();
        List<T> rev = new ArrayList<>(n);
        for (int i = n - 1; i >= 0; i--) {
            rev.add(vc.get(i));
        }
        return rev;
    }

    public static void main(String[] args) {
        List<Integer> vc = new ArrayList<>() ;
        vc.add(1);
        vc.add(2);
        vc.add(3);
        
        List<Integer> rev = reverseCopy(vc);
        for (int i : rev) {
            System.out.print(i + " ");
        }
        System.out.println();
    }
}

// No builtin function to reverse an ArrayList and return a copy

